FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,124 @@
+
+### 2.2 Explicit vs. implicit index creation
+
+#### 2.2.1 Dataset/DataArray's `set_index` method
+
+New indexes can be built from an existing set of coordinates or variables in a Dataset/DataArray using the `.set_index()` method.
+
+TODO: describe API updates needed to provide the kind of index that we want to build from a set of coordinates.
+
+#### 2.2.2 Dataset/DataArray's `indexes` constructor argument
+
+TODO
+
+#### 2.2.3 Implicit default indexes
+
+In general explicit index creation should be preferred over implicit index creation. However, there is a majority of cases where basic `pandas.Index` objects could be built and used as indexes for 1-dimensional coordinates. For convenience, Xarray should automatically build such indexes for the coordinates where no index has been explicitly assigned in the Dataset/DataArray constructor or when indexes have been reset / dropped.
+
+For which coordinates?
+
+- A. only 1D coordinates with a name matching their dimension name
+- B. all 1D coordinates
+
+When to create it?
+
+- A. each time when a new Dataset/DataArray is created
+- B. only when we need it (i.e., when calling `.sel()` or `indexes`)
+
+#### 2.2.4 Implicit coordinates
+
+If we want to strictly follow the data model defined in [Section 1](#1_Data_Model), we need to ensure that at least one coordinate exists for each index (generally we need to ensure that the data indexed is properly exposed as one or more coordinates), especially when index objects are given directly to the Dataset/DataArray constructor ([Section 2.2.2](#222_DatasetDataArrays_indexes_constructor_argument)). This means that every `XarrayIndex` subclass must implement a method to get or generate one or more Xarray coordinates.
+
+Additionally, we might want to generate a coordinate that is specific to the index, generally with a composite data type, like a coordinate with n-elements tuple labels for a n-levels `pandas.MultiIndex`. This should be optional, though. An `XarrayIndex` may or may not support this. This coordinate probably shouldn't be created implicitly or by default when setting the new index. Also, this might raise some issues, e.g., what would be the name of this coordinate? For a `pandas.MultiIndex` it would make sense in many (but not all) cases to choose the dimension name as the coordinate name.
+
+### 2.3 Index access
+
+#### 2.3.1 Dataset/DataArray's `indexes` property
+
+The `indexes` property would return a mapping, where keys are coordinate name(s) (i.e., either a string or a tuple) and values are index objects (i.e., instances of `XarrayIndex`). If a tuple is used as key, the order of its elements should be consistent with the order in which each coordinate participate to the index (see [Section 1](#1_Data_Model)).
```

Comment:
What do we need multiple names in the key for? Are there cases where we want to look-up different indexes depending on which variables are being used?

Rather than allowing for tuple keys, I was imagining we could allow for multiple coordinate names to point to the same index. This would support mapping multiple coordinates to the same (typically multi-dimensional) index, e.g., both `'lat'` and `'lon'` would map to the same index. I think this would align well with the "one-to-zero-or-one" relationship of "Coordinate -> Index" that you describe above.

Reply:
> Right now, it is acceptable to have a tuple as dimension name, so this would be a breaking change

Note that the keys of the mapping returned by the `indexes` property would not correspond here to dimension names but rather one or more coordinate names.

>  Rather than allowing for tuple keys, I was imagining we could allow for multiple coordinate names to point to the same index.

Yes I saw that this is what you suggest in https://github.com/pydata/xarray/pull/2195#issuecomment-442554179. However, I was rather thinking here about something that we could use to quickly iterate over all the indexes in a Dataset/DataArray. Having a collection of unique index objects as values seems better adapted for this use case IMO. It would also align well with the API of `set_index` or the DataArray/Dataset constructors (the `indexes` argument) where we'll probably need to provide some coordinate/variable name(s) -> index class/object as keys -> values. Either way, the structure returned by `indexes` should be consistent with the DataArray/Dataset repr depending on how we show the indexes (I'm not sure that the best repr would be to show an index section with duplicated lines for multi-coordinate indexes, although I don't have a clear idea / strong opinion on this yet).

With what you suggest in https://github.com/pydata/xarray/pull/4979#discussion_r585197016, we could even imagine returning a list of `XarrayIndex` objects, but I think returning a mapping is slightly more convenient as it doesn't hide the index / coordinate(s) relationship behind the index object.

That said, I agree that we need something for getting an index given some coordinate. This is why I suggest another property or method in Section 2.3.2.

Reply:
For the repr, we could probably consolidate entries that map to the same index object, e.g.,
```
Indexes:
  lat, lon     <SpatialIndex coords=(lat, lon) dims=(x, y)>
  x            Float64Index
  y            Float64Index
```

Given that the `repr` prints all the indexes, it is probably OK to group them even if the indexes map has separate entries for `lat` and `lon`. There would just be an extra validation step when you add/remove an index that verifies that the index is consistent on all matching coordinates.

It might be worth sketching out a brief list of trade-offs with each representation for `indexes`, e.g.,
- `Dict[CoordinateName, XarrayIndex]` with duplicated values: Easy to look up all matching indexes by coordinate name, even if the index is over multiple coordinates. Iterating over _unique_ indexes requires some care (e.g., a separate `set` to keep track of Index `id()`).
- `Dict[Tuple[CoordinateName, ...], XarrayIndex]` with unique values: Easy to look-up matching indexes if all coordinate names are supplied in the correct order (but I think this is relatively rare). Cases like `da.sel(lon=..., lat=...)` (shared index, wrong order) or `da.sel(x=..., time=...)` (separate indexes) would require iterating over all indexes and building some sort of temporary structure. Also, there's potential ambiguity between names and tuples of names (if we also allow raw coordinate names in the dict).

Reply:
I think that you convinced me that there's probably more practical reasons to choose `Dict[CoordinateName, XarrayIndex]` as return type for `indexes`.